keps/sig-storage/20200212-volume-group.md

```diff
@@ -0,0 +1,1445 @@
+	......
+        // +optional
+        VolumeGroupSnapshotName *string
+	......
+}
+
+type VolumeSnapshotContentStatus struct{
+	......
+        // +optional
+        VolumeGroupSnapshotContentName *string
+	......
+}
+```
+
+### Example Yaml Files
+
+#### Create Volume Group
+
+Example yaml files to create a VolumeGroupClass and a VolumeGroup are in the following.
+
+Create a VolumeGroupClass that supports volumeGroupSnapshot:
+```
+apiVersion: volumegroup.storage.k8s.io/v1alpha1
+kind: VolumeGroupClass
+metadata:
+  name: volumeGroupClass1
+spec:
+  parameters:
+     …...
+ supportVolumeGroupSnapshot: true
+```
+
+Create a VolumeGroup belongs to this VolumeGroupClass:
+```
+apiVersion: volumegroup.storage.k8s.io/v1alpha1
+kind: VolumeGroup
+metadata:
+  Name: volumeGroup1
+spec:
+  volumeGroupClassName: volumeGroupClass1
+```
+
+#### Add PVC to VolumeGroup
+
+Create a PVC that belongs to the volume group which supports volumeGroupSnapshot:
+```
+apiVersion: v1
+kind: PersistentVolumeClaim
+metadata:
+  name: pvc1
+spec:
+  accessModes:
+  - ReadWriteOnce
+  dataSource: null
+  resources:
+	requests:
+  	storage: 1Gi
+  storageClassName: storageClass1
+  volumeMode: Filesystem
+  volumeGroupNames: [volumeGroup1]
+```
+
+#### Create VolumeGroupSnapshot
+
+Create a VolumeGroupSnapshotClass:
+```
+apiVersion: volumegroup.storage.k8s.io/v1alpha1
+kind: VolumeGroupSnapshotClass
+metadata:
+  name: volumeGroupSnapshotClass1
+spec:
+  parameters:
+     …...
+```
+
+A VolumeGroupSnapshot taken from the VolumeGroup dynamically:
+```
+apiVersion: volumegroup.storage.k8s.io/v1alpha1
+kind: VolumeGroupSnapshot
+metadata:
+  name: my-group-snapshot
+spec:
+  source:
+      volumeGroupName: volumeGroup1
+  volumeGroupSnapshotClassName: volumeGroupSnapshotClass1
+```
+
+A new external VolumeGroup controller will handle VolumeGroupClass, VolumeGroup, and VolumeGroupContent resources. We may need to split this into two controllers, one common controller that handles common functions such as binding, and one sidecar controller that calls the CSI driver.
+
+External provisioner will be modified to read information from volume groups (through volumeGroupNames) and pass them down to the CSI driver.
+
+A new external VolumeGroupSnapshot controller will handle VolumeGroupSnapshotClass, VolumeGroupSnapshot, and VolumeGroupSnapshotContent resources. We may need to split this into two controllers, one common controller that handles common functions such as binding, and one sidecar controller that calls the CSI driver.
+
+Snapshot controller will be modified to update VolumeSnapshot status. External snapshotter sidecar will be modified to update VolumeSnapshotContent status.
+
+### CSI Changes
+
+#### CSI Capabilities
+
+New controller capabilities VOLUME_GROUP, VOLUME_GROUP_ADD_REMOVE_EXISTING_VOLUME, GROUP_SNAPSHOT, INDIVIDUAL_SNAPSHOT_RESTORE, GET_VOLUME_GROUP, GET_VOLUME_GROUP_SNAPSHOT, LIST_VOLUME_GROUPS, LIST_VOLUME_GROUP_SNAPSHOTS will be added.
+
+* VOLUME_GROUP:
```

Comment:
Change to CREATE_DELETE_VOLUME_GROUP to be consistent with other similar capabilities.